sonar-xml-plugin/src/main/java/org/sonar/plugins/xml/highlighting/XMLHighlighting.java

```diff
@@ -0,0 +1,216 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.sonar.plugins.xml.highlighting;
+
+import com.google.common.io.Files;
+
+import javax.xml.stream.Location;
+import javax.xml.stream.XMLInputFactory;
+import javax.xml.stream.XMLStreamConstants;
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.XMLStreamReader;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.InputStream;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+
+public class XMLHighlighting {
+
+  private static final String XML_DECLARATION_TAG = "<?xml";
+
+  private List<HighlightingData> highlighting = new ArrayList<>();
+  private String content;
+
+  public XMLHighlighting(File file, Charset charset) throws Exception{
+    content = Files.toString(file, charset);
+    highlightXML(new FileInputStream(file));
+  }
+
+  public XMLHighlighting(String xmlStrContent) throws XMLStreamException{
+    content = xmlStrContent;
+    InputStream inputStream = new ByteArrayInputStream(xmlStrContent.getBytes(StandardCharsets.UTF_8));
+    highlightXML(inputStream);
+  }
+
+  public List<HighlightingData> getHighlightingData() {
+    return highlighting;
+  }
+
+  public void highlightXML(InputStream inputStream) throws XMLStreamException {
+    XMLInputFactory factory = XMLInputFactory.newInstance();
+    factory.setProperty(XMLInputFactory.SUPPORT_DTD, "false");
+    XMLStreamReader xmlReader = factory.createXMLStreamReader(inputStream);
+    highlightXmlDeclaration();
+
+    while (xmlReader.hasNext()) {
+      Location prevLocation = xmlReader.getLocation();
+      xmlReader.next();
+      int startOffset = xmlReader.getLocation().getCharacterOffset();
+
+      switch (xmlReader.getEventType()) {
+        case XMLStreamConstants.START_ELEMENT:
+          highlightStartElement(xmlReader, startOffset);
+          break;
+
+        case XMLStreamConstants.END_ELEMENT:
+          highlightEndElement(xmlReader, prevLocation, startOffset);
+          break;
+
+        case XMLStreamConstants.CDATA:
+          highlightCData(startOffset);
+          break;
+
+        case XMLStreamConstants.DTD:
+          highlightDTD(startOffset);
+          break;
+
+        case XMLStreamConstants.COMMENT:
+          // 7 is length of opening and closing symbols : "<!--" and "-->"
+          highlighting.add(new HighlightingData(startOffset, startOffset + xmlReader.getTextLength() + 7, "j"));
+          break;
+
+        default:
+          break;
+      }
+    }
+  }
+
+  private void highlightDTD( int startOffset) {
+    int closingBracketStartOffset;
+    closingBracketStartOffset = getTagClosingBracketStartOffset(startOffset);
```

Comment:
@pynicolas I see your point, but still it isn't trivial. We need data from next events while processing current one. I've spent about hour to do that, but still some tests are failing. I let's keep it as it is.